src/base/bittorrent/torrentimpl.cpp

```diff
@@ -2037,9 +2037,29 @@ void TorrentImpl::handleFileCompletedAlert(const lt::file_completed_alert *p)
         return;
 
     const int fileIndex = m_indexMap.value(p->index, -1);
-    Q_ASSERT(fileIndex >= 0);
+    Q_ASSERT((fileIndex >= 0) && (fileIndex < filesCount()));
+    if ((fileIndex < 0) || (fileIndex >= filesCount()))
+    {
+        LogMsg(tr("Logical error encountered. Please report to developers. Torrent: '%1'."
+                  " Metadata exists: %2. Native file index: %3. File index: %4.")
+               .arg(name(), (hasMetadata() ? tr("Yes") : tr("No")), QString::number(static_cast<int>(p->index)), QString::number(fileIndex))
+               , Log::CRITICAL);
+        return;
+    }
 
-    m_completedFiles[fileIndex] = true;
+    try
+    {
+        m_completedFiles[fileIndex] = true;
```

Comment:
According to the text in the description of [QBitArray](https://doc.qt.io/qt-6/qbitarray.html#details) it is faster to do setBit()/testBit() than using the [] operator.
So maybe convert to
```c++
m_completedFiles.setBit(fileIndex, true);
```

???

Reply:
> In case you aren't inclined to change this, let me know so I'll approve the PR.

I'm sorry if that confused you. This PR was published for testing purposes so that the affected user could use its build artifact and provide additional diagnostic information about the error.

P.S. Your proposal can be applied in some productive PR.